src/cargo/core/compiler/mod.rs

```diff
@@ -631,6 +605,10 @@ fn rustdoc<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoResult
         rustdoc.arg("--cfg").arg(&format!("feature=\"{}\"", feat));
     }
 
+    if bcx.build_config.json_messages() {
+        rustdoc.arg("--error-format").arg("json");
```

Comment:
We don't need to feature-detect support for `json` here, right? The user requests `--message-format=json` explicitly, so they'll see an error and that is expected behavior.

Reply:
I don't understand the question.  This is the code that translates cargo's `--message-format` flag to rustdoc's `--error-format` flag.

Reply:
`--error-format` is a recent addition to `rustdoc`, so, in theory, using new cargo and old rustdoc could result in a 

```
error: the `-Z unstable-options` flag must also be passed to enable the flag `error-format`
```

message. However, that should be fine, because the user has to ask for `cargo doc --message-format=json` explicitly. That is, it's not a flag that we add automatically for existing workflows, so we don't have to feature-detect support for it.

Reply:
Oh, I understand now.  Yea, my understanding is that the version of cargo is tightly wedded to rustc and rustdoc.  I think some distributions have attempted to split them, but afaik cargo doesn't really support that use case, right?  At least in this case, it wouldn't break the normal usage.

Reply:
It is at least somewhat supported: you can override compiler & rustdoc with env vars, and we try account for different compilers in Rustc cache. I think the current approach is "don't break things without necessity", and I don't remember any actual CLI changes that could have caused breakage.